### ISSUE TITLE
CHECK constraint lost during upgrade to v19.1.0

### ISSUE THREAD

**Issue by charsleysa** [cockroachdb/cockroach#37285]
**Describe the problem**

Please describe the issue you observed, and any steps we can take to reproduce it:
CHECK constraints on tables are being lost when upgrading from v2.1.5/6 to v19.1.0

**To Reproduce**

What did you do? Describe in your own words.

If possible, provide steps to reproduce the behavior:

1. Create database with CRDB v2.1.6
2. Create table with a CHECK constraint
3. Upgrade CRDB to v19.1.0
4. Run sql `show constraints from <table_name>;`

**Expected behavior**
Nothing, the CHECK constraints should be untouched.

**Additional data / screenshots**
Not available.

**Environment:**
 - CockroachDB version [e.g. 2.0.x] - v19.1.0
 - Server OS: [e.g. Linux/Distrib] - Amazon Linux 2.0 / Docker (when testing locally)
 - Client app [e.g. `cockroach sql`, JDBC, ...] - DataGrip / Node.js pg

**Additional context**
What was the impact?
Lost all CHECK constraints on existing tables.
There was no warning that CHECK constraints were dropped, nothing in the Web UI nor in the logs from what I can see.

**Comment by charsleysa** [comment]
UPDATE:
What makes this even weirder is that the CHECK constraints aren't truly lost!

When I try to insert a row that violates the CHECK constraint I still get an error saying that I have violated the CHECK constraint, but I am unable to see the constraint using `show constraints from <table_name>;` nor am I able to drop the constraint using `ALTER TABLE <table_name> DROP CONSTRAINT <constraint_name>;`.

The only workaround I have found is to create the CHECK constraints again, __DROP__ the CHECK constraints, then create them again. This seems to clear out the old lost constraints.

**Comment by jordanlewis** [comment]
Hey @dt, would you be able to investigate this issue? Thanks!

**Comment by roncrdb** [comment]
Hey @charsleysa ,

I was unable to reproduce this issue: 

Here is a sample output of my steps, please let me know if there is something I missed. 
```
Version 2.1.6

# Create a new database, created a new table, create a new check, insert to validate check. 
CREATE DATABASE test;
USE test; 
create table checktest (a string check (a ~ '^[A-Za-z]$'));
insert into checktest values ('é');

Error: pq: failed to satisfy CHECK constraint (a ~ '^[A-Za-z]$')


Upgraded to 19.1

Opened the SQL client

 Server version: CockroachDB CCL v19.1.0 (x86_64-unknown-linux-gnu, built 2019/04/29 18:36:40, go1.11.6) (same version as client)
# Cluster ID: 9c28722e-19a8-4941-a114-f546c4cc121e
# Organization: Cockroach Labs - Production Testing
#
# Enter \? for a brief introduction.
#
root@localhost:26257/defaultdb> use test;
SET

Time: 498.478µs

root@localhost:26257/test> show constraints from checktest;
  table_name | constraint_name | constraint_type |         details          | validated
+------------+-----------------+-----------------+--------------------------+-----------+
  checktest  | check_a         | CHECK           | CHECK (a ~ '^[A-Za-z]$') |   true
(1 row)

Time: 6.544894ms

root@localhost:26257/test> insert into checktest values ('é');
pq: failed to satisfy CHECK constraint (a ~ '^[A-Za-z]$')
```

**Comment by charsleysa** [comment]
Hi @roncrdb 

I have created a repro but it's slightly more convoluted than your attempt since I've trimmed it down from our system setup. I've added the required scripts below so it should be mostly automated, I've also added logs from when I ran it on my machine.

Steps:
1) Ensure CockroachDB is installed on host machine so that it can be used to connect to db (I used v2.1.4 which was already on my dev machine)
2) run `docker-compose up -d` in the directory where the `docker-compose.yml` file is located (note: the setup utilises port 26257 and 8080 on the host machine)
3) connect to db using `cockroach sql --insecure`
4) copy and paste contents of `setup_database.sql` and execute, all commands should execute successfully except that last one which is an insert with a value that violates the check constraint and is expect to throw an error
5) update the `docker-compose.yml` file and replace `v2.1.6` with `v19.1.0`
6) re-run `docker-compose up -d` so that the db is upgraded
7) re-connect to db using `cockroach sql --insecure`
8) copy and paste contents of `test_database.sql` and execute

Expected Result:
The commands from `test_database.sql` should show an empty constraints table followed by a failed insert due to a violation of a check constraint (which does not appear in results from the show constraints command)

<details>
<summary>pre-upgrade.log</summary>

```
Stefans-MBP:crdb-check-issue stefan$ docker-compose up -d
Creating network "crdb-checktest" with the default driver
Creating volume "crdb-check-issue_crdb01-checktest-data" with local driver
Creating volume "crdb-check-issue_crdb02-checktest-data" with local driver
Creating volume "crdb-check-issue_crdb03-checktest-data" with local driver
Pulling crdb01-checktest (cockroachdb/cockroach:v2.1.6)...
v2.1.6: Pulling from cockroachdb/cockroach
85b1f47fba49: Already exists
3d0fb0094e96: Pull complete
8f3fdbdbe1ca: Pull complete
acf7d8723bed: Pull complete
Creating crdb01-checktest ... done
Creating crdb03-checktest ... done
Creating crdb02-checktest ... done
Stefans-MBP:crdb-check-issue stefan$ cockroach sql --insecure
# Welcome to the cockroach SQL interface.
# All statements must be terminated by a semicolon.
# To exit: CTRL + D.
#
# Client version: CockroachDB CCL v2.1.4 (x86_64-apple-darwin18.2.0, built 2019/01/21 22:57:36, go1.11.4)
# Server version: CockroachDB CCL v2.1.6 (x86_64-unknown-linux-gnu, built 2019/03/04 23:21:07, go1.10.7)
# Cluster ID: 5d8a7f45-8f36-4a4a-bb2d-04efb89f77d5
#
# Enter \? for a brief introduction.
#
root@:26257/defaultdb> CREATE USER techadmin;                                                                                                                                                                                                                                                       CREATE DATABASE checktest WITH ENCODING = 'UTF-8';                                                                                                                                                                                                                                                  GRANT ALL ON DATABASE checktest TO techadmin;                                                                                                                                                                                                                                                       USE checktest;                                                                                                                                                                                                                                                                                      CREATE TABLE checktest(testcolumn STRING NOT NULL);                                                                                                                                                                                                                                                 ALTER TABLE checktest ADD CONSTRAINT check_checktest_testcolumn CHECK (testcolumn IN ('CHECK123', 'CHECK456'));                                                                                                                                                                                     INSERT INTO checktest VALUES ('CHECKINVALID');                                                                                                     
CREATE USER 1

Time: 61.995ms

CREATE DATABASE

Time: 45.562ms

GRANT

Time: 40.319ms

SET

Time: 2.689ms

CREATE TABLE

Time: 45.627ms

ALTER TABLE

Time: 125.086ms

pq: failed to satisfy CHECK constraint (testcolumn IN ('CHECK123', 'CHECK456'))
root@:26257/checktest> 
```

</details>

<details>
<summary>post-upgrade.log</summary>

```
Stefans-MBP:crdb-check-issue stefan$ docker-compose up -d
Recreating crdb02-checktest ... done
Recreating crdb03-checktest ... done
Recreating crdb01-checktest ... done
Stefans-MBP:crdb-check-issue stefan$ cockroach sql --insecure
# Welcome to the cockroach SQL interface.
# All statements must be terminated by a semicolon.
# To exit: CTRL + D.
#
# Client version: CockroachDB CCL v2.1.4 (x86_64-apple-darwin18.2.0, built 2019/01/21 22:57:36, go1.11.4)
# Server version: CockroachDB CCL v19.1.0 (x86_64-unknown-linux-gnu, built 2019/04/29 18:36:40, go1.11.6)
# Cluster ID: 5d8a7f45-8f36-4a4a-bb2d-04efb89f77d5
#
# Enter \? for a brief introduction.
#
root@:26257/defaultdb> USE checktest;                                                                                                                                                                                                                                                               SHOW CONSTRAINTS FROM checktest;                                                                                                                                                                                                                                                                    INSERT INTO checktest VALUES ('CHECKINVALID');                                                                                                     
SET

Time: 1.603ms

  table_name | constraint_name | constraint_type | details | validated  
+------------+-----------------+-----------------+---------+-----------+
(0 rows)

Time: 18.176ms

pq: failed to satisfy CHECK constraint (testcolumn IN ('CHECK123':::STRING, 'CHECK456':::STRING))
root@:26257/checktest> 
```

</details>

<details>
<summary>docker-compose.yml</summary>

```yaml
version: '3.6'

networks:
  crdb-checktest:
    name: crdb-checktest

volumes:
  crdb01-checktest-data:
    driver: local
  crdb02-checktest-data:
    driver: local
  crdb03-checktest-data:
    driver: local

services:
  crdb01-checktest:
    container_name: crdb01-checktest
    image: 'cockroachdb/cockroach:v19.1.0'
    ports:
    - 26257:26257
    - 8080:8080
    volumes:
    - 'crdb01-checktest-data:/cockroach/cockroach-data'
    networks:
      crdb-checktest:
        aliases:
        - crdb01-checktest
    command: ['start', '--insecure']

  crdb02-checktest:
    container_name: crdb02-checktest
    image: 'cockroachdb/cockroach:v19.1.0'
    volumes:
    - 'crdb02-checktest-data:/cockroach/cockroach-data'
    networks:
      crdb-checktest:
        aliases:
        - crdb02-checktest
    command: ['start', '--insecure', '--join=crdb01-checktest']

  crdb03-checktest:
    container_name: crdb03-checktest
    image: 'cockroachdb/cockroach:v19.1.0'
    volumes:
    - 'crdb03-checktest-data:/cockroach/cockroach-data'
    networks:
      crdb-checktest:
        aliases:
        - crdb03-checktest
    command: ['start', '--insecure', '--join=crdb01-checktest']
```

</details>

<details>
<summary>setup_database.sql</summary>

```sql
CREATE USER techadmin;

CREATE DATABASE checktest WITH ENCODING = 'UTF-8';

GRANT ALL ON DATABASE checktest TO techadmin;

USE checktest;

CREATE TABLE checktest(testcolumn STRING NOT NULL);

ALTER TABLE checktest ADD CONSTRAINT check_checktest_testcolumn CHECK (testcolumn IN ('CHECK123', 'CHECK456'));

INSERT INTO checktest VALUES ('CHECKINVALID');

```

</details>

<details>
<summary>test_database.sql</summary>

```sql
USE checktest;

SHOW CONSTRAINTS FROM checktest;

INSERT INTO checktest VALUES ('CHECKINVALID');

```

</details>

**Comment by lucy-zhang** [comment]
I was able to reproduce this. The problem is that `AllActiveAndInactiveChecks()` (https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/sqlbase/structured.go#L587-L605) erroneously ignores all non-validated check constraints, since it assumes that all check constraints will be created in a validated state in 19.1.

As you've found, the check constraint still exists on the table descriptor and is still enforced for all writes, but it's not visible to `SHOW CONSTRAINTS` or `ALTER TABLE` (hence not able to be dropped, or validated with `VALIDATE CONSTRAINT`). I'm working on a fix.

**Comment by lucy-zhang** [comment]
This only happens if the check constraint was originally added in an `ALTER TABLE` statement and then never validated with `ALTER TABLE ... VALIDATE CONSTRAINT`. If the check constraint was in the original `CREATE TABLE` statement, it starts off in the validated state and is not affected by this bug, which is why @roncrdb's example works fine.

**Comment by lucy-zhang** [comment]
The fix is getting backported to 19.1: #37462
